dbms/src/Storages/DeltaMerge/Index/InvertedIndex/Writer.cpp

```diff
@@ -0,0 +1,280 @@
+// Copyright 2025 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Columns/ColumnNullable.h>
+#include <Common/Stopwatch.h>
+#include <Common/TiFlashMetrics.h>
+#include <Functions/FunctionHelpers.h>
+#include <IO/Buffer/WriteBufferFromFile.h>
+#include <Storages/DeltaMerge/Index/InvertedIndex/Writer.h>
+
+#include <ext/scope_guard.h>
+
+namespace DB::ErrorCodes
+{
+extern const int ABORTED;
+extern const int BAD_ARGUMENTS;
+} // namespace DB::ErrorCodes
+
+namespace DB::DM
+{
+
+template <typename T>
+void InvertedIndexWriterInternal<T>::addBlock(
+    const IColumn & column,
+    const ColumnVector<UInt8> * del_mark,
+    ProceedCheckFn should_proceed)
+{
+    // Note: column may be nullable.
+    const bool is_nullable = column.isColumnNullable();
+    const auto * col_vector
+        = is_nullable ? checkAndGetNestedColumn<ColumnVector<T>>(&column) : checkAndGetColumn<ColumnVector<T>>(&column);
+    RUNTIME_CHECK_MSG(col_vector, "ColumnVector is expected, get: {}, T: {}", column.getName(), typeid(T).name());
+    const auto & col_data = col_vector->getData();
+
+    const auto * null_map = is_nullable ? &(checkAndGetColumn<ColumnNullable>(&column)->getNullMapData()) : nullptr;
+    const auto * del_mark_data = del_mark ? &(del_mark->getData()) : nullptr;
```

Comment:
Why is it possible that del_mark is null?

Reply:
in test, we will pass `nullptr`